src/server/pkg/collection/iterator.go

```diff
@@ -0,0 +1,190 @@
+package collection
+
+import (
+	"fmt"
+	"sort"
+	"sync/atomic"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type Options struct {
+	Target   etcd.SortTarget
+	Order    etcd.SortOrder
+	SelfSort bool
+}
+
+var DefaultOptions *Options = &Options{etcd.SortByCreateRevision, etcd.SortDescend, false}
+
+type etcdIterator interface {
+	next() ([]*mvccpb.KeyValue, error)
+	done() bool
+}
```

Comment:
One of the goals I had for this PR was to remove this iterator pattern from the code base, at least from the collections library and other parts that I'm touching in this PR. After using this pattern I'm convinced that it's not a very good way to do iteration in go for a couple of reasons:

1. It almost always leads to longer code, you can see this if you look above, you have code that reads:

```
iter := newEtcdIterator(c, prefix, limitPtr, opts)
for !iter.done() {
    kvs, err := iter.next()
    if err != nil {
	return err
    }
    // real code begins here
}
```

If this was using a callback it would look like:

```
return iterate(c, prefix,limitPtr, opts, func(kvs *mvccpb.KeyValues) error {
    // real code begins here
})
```

2 lines of boilerplate vs 8, this stuff really starts to add up too because anytime you want to compose one iterator to make another you need to write another for-loop. I probably cleaned up several hundred lines of iteration boilerplate as part of this PR.

2. There's a lot more reasonable ways to do iterator interfaces and as a result you have to read each individual iterator to see how it was implemented. In particular: some interfaces opt give you method to figure out if the iteration is complete, you've called yours done and it's true if iteration has finished, but it often makes more sense to have one that's true if iteration has not finished It's also common to have neither method but to have your `next` method return `ok` as it second return value if you can continue. Or sometimes it just returns an error and if you're at the end of the iteration that error is `io.EOF`. With a callback there's only one reasonable way to implement it.

3. There are certain optimizations that iterators make much harder. In particular, a really nice optimization for this code would be to asynchronously request the next batch will the first batch is being processed. To do that with iterators you would need to have a goroutine running in the background that you launch from you `next`, but now that goroutine could potentially leak, so you need to add a cancel or close method to your iterator, and callers of the iterator need to remember to defer that method everytime they create an iterator. Which is more boilerplate and potential bugs.

So long story short, let's migrate this over to callback interface like the other iteration methods in this PR and I'm going to add these thoughts on iteration to our official coding guidelines. I was actually planning to do that anyways as part of this PR.

Reply:
This makes a lot of sense. I will go ahead and start migrating this over to a callback interface and make a mental note for when I add/change things in the future.